Add tests+fix for an incorrectly case of generated inlined C code from CCodeGenerator

hernanmd opened the pull request:
When the code in written to C from the CAST, a complex expression from an inlining case could be adding one variable to a anInteger. But this is invalid when the expression is found in a L-Value. In this case, the sum operation should be enclosed between parentheses (both TAST and CAST are not affected as they're correctly generated). For example in:
	
```c
	(pTo + anInteger)[i] = pFrom + anInteger[i]
```	

instead of (without parentheses)
	
```c
	pTo + anInteger[i] = pFrom + anInteger[i]
```

This PR includes a fix implemented in `CCodeGenerator>>generateCASTAtPut:` to add parentheses to the l-value.
It also includes test handling cases of annotated (with pragmas) and unannotated variables for inlinings.
Similarily, tests cases were added for generation of prepared (with prepareMethods) and unprepared methods are considered.

Commit:
Add tests for an incorrectly case of generated inlined C code from the CCodeGenerator.

guillep commented on smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:
```diff
@@ -1155,6 +1155,163 @@ SlangBasicTranslationTest >> testGoto [
+	| tMethod translation |
+	tMethod := (self getTMethodFrom: #methodSumParameters:to:with:).
+	generator 
+		prepareMethods;
```
I think preparing methods should be mandatory, as Slang does it all the time when generating the VM.
I don't think Slang is prepared to translate something if this is not done.

Actually, what we would like to test is:
 - a test where the operand is inlined in the array access `(a + b)[i] = ...` => this we have it
 - a test where the array pointer is outlined into a variable `var := expression; expression[i] = ...` => this we can do it by having a piece of code that cannot be inlined. E.g., a function call!

hernanmd replied:
Ok. I've updated the PR with the suggested changes.

guillep commented on smalltalksrc/Slang/CCodeGenerator.class.st:
```diff
@@ -1534,10 +1534,18 @@ CCodeGenerator >> generateCASTAt: tast [
 	^ CAssignmentNode
 		  lvalue: (CArrayAccessNode
-				   array: (tast receiver asCASTExpressionIn: self)
+				   array: ((tast receiver asCASTExpressionIn: self) needsParentheses: true)
```
👍

guillep commented on smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:
```diff
@@ -1155,6 +1155,163 @@ SlangBasicTranslationTest >> testGoto [
+	pFrom1 = pFrom + anInteger;
+	pTo1 = pTo + anInteger;
+	for (i = 0; i <= 5; i += 1) {
+		pTo1[i] = pFrom1[i];
```
Does this compile? I think it does not, right? see https://godbolt.org/z/E4feMdv6E
=> it is testing the wrong behavior, and I don't like committing to wrong :).

I think we should remove this test.

We could fix it in a follow-up if needed

hernanmd replied:
Yes, I would remove them but I'd like to check later, because what I meant is something like this: https://godbolt.org/z/jn97EaP96

So right now the inlining logic does not "pass" the pragma types to the signature?

Commit:
Changed tests as suggested to check Slang methods annotated with types, either using built-in and without built-in C operations.
Simplify test cases using getTMethod: by automatically preparing the method.
Removed test cases where invalid C were generated.

guillep committed:
Merge branch 'pharo-12' into test_inlinings_1

guillep reviewed (approved):
Tests are green after changes and merge!

Thanks @hernanmd !

PalumboN commented on smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:
```diff
@@ -1176,6 +1178,74 @@ SlangBasicTranslationTest >> testGoto [
+	self 
+		assert: translation
+		equals: '/* SlangBasicTranslationTestClass>>#methodUseParametersWithAnnotationsBuiltIn:to:with: */
+static sqInt
+methodUseParametersWithAnnotationsBuiltIntowith(unsigned int *pFrom, unsigned int *pTo, sqInt anInteger)
+{
+	sqInt i;
+	sqInt pFrom1;
+	sqInt pTo1;
+
+	/* begin methodFromWithAnnotations:to: */
+	pFrom1 = ((sqInt) (pFrom + anInteger) );
+	for (i = 0; i <= 10; i += 1) {
+		pTo[i] = pFrom1[i];
+	};
+	return 0;
+}'
```
This `pFrom1[i]` compiles in C? Using `sqInt` as type?

PalumboN replied:
And why `sqInt pTo1` is defined? It's never used. Could be a bug with the renames?

guillep replied:
True! I wonder where the cast to `sqInt` comes from? I think it's the inlining that loses the parameter type.



> And why `sqInt pTo1` is defined? It's never used. Could be a bug with the renames?

That's an inlining problem, it's not specific to this.

PalumboN commented on smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:
```diff
@@ -1176,6 +1178,74 @@ SlangBasicTranslationTest >> testGoto [
+	self
+		assert: translation
+		equals: '/* SlangBasicTranslationTestClass>>#methodUseParametersWithAnnotations:to:with: */
+static sqInt
+methodUseParametersWithAnnotationstowith(unsigned int *pFrom, unsigned int *pTo, sqInt anInteger)
+{
+	sqInt i;
+	sqInt pFrom1;
+	unsigned int *pTo1;
+
+	/* begin methodFromWithAnnotations:to:len: */
+	pTo1 = ((unsigned int *) (yourself(pTo)) );
+	for (i = 0; i < 5; i += 1) {
+		pTo1[i] = pFrom + anInteger[i];
+	};
+	return 0;
+}'
+]
```
`anInteger[i]` is valid C?

guillep replied:
Good catch. Actually, here the parentheses are missing on the right hand side too. It should be

```c
 		pTo1[i] = (pFrom + anInteger)[i];
```

PalumboN commented on smalltalksrc/Slang/CCodeGenerator.class.st:
```diff
@@ -1538,10 +1538,18 @@ CCodeGenerator >> generateCASTAt: tast [
 
 { #category : #'CAST translation' }
 CCodeGenerator >> generateCASTAtPut: tast [
+	" Add parentheses to lvalue to correctly generate C code for complex inlining cases. For example in:
+	
+	 '(pTo + anInteger)[i] = pFrom + anInteger[i]'
+	
+	instead of (without parentheses)
+	
+	'pTo + anInteger[i] = pFrom + anInteger[i]'
```
I understand that this is the goal of this PR, but this case is not present in the new tests 😞

PalumboN replied:
And here also says `anInteger[i]` which is a weird example.

guillep replied:
Yes parentheses should appear on the right too. We should fix it in a separate PR

PalumboN reviewed (commented):
I saw the discussion but I think that I missed something. I left some questions.

guillep reviewed (commented):
@PalumboN I answered your questions.
This PR actually was meant to fix the issue we spotted in the assignment's left-hand side.

Now we spotted more issues thanks to you :)
I'll open issues for them
